### PR TITLE
[MTDSA-5534] - Amend Relief Investments Request Parser & Validator

### PR DESCRIPTION
**Pull request by DreamcastKnight**

### CODE REVIEW

File: app/v1/controllers/requestParsers/validators/AmendReliefInvestmentValidator.scala [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+        case Nil =>
+          Nil
+        case paths =>
+          List(FormatInvestmentRefErrorGenerator.generate(paths))
+      }
+
+    }
+
+    List(
+      // all other errors you'd do on these things I guess
```
Comment (james-work-account) [inline app/v1/controllers/requestParsers/validators/AmendReliefInvestmentValidator.scala:140]: These can be taken out, they were more notes than anything IIRC

File: app/v1/controllers/requestParsers/validators/validations/InvestmentRefValidation.scala [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+ * limitations under the License.
+ */
+
+package v1.controllers.requestParsers.validators.validations
+
+
+object InvestmentRefValidation {
+
+  private val investRegex =
+    "^[A-Za-z0-9/]{1,}+$"
```
[the remark this answers is not in the document]
Comment (james-work-account) [reply app/v1/controllers/requestParsers/validators/validations/InvestmentRefValidation.scala:23]: Also, we're not allowing slashes in the ref (according to the spec) - is the spec wrong?

Also also, `{1,}` and `+` mean exactly the same thing (ie, `one or more times`) so the `{1,}` is redundant

[the remark this answers is not in the document]
Comment (DreamcastKnight) [reply app/v1/controllers/requestParsers/validators/validations/InvestmentRefValidation.scala:23]: I think the spec potentially needs rectifying, as some of the unique investment ref's have a / in them, so I needed to add it in the regex

File: app/v1/controllers/requestParsers/validators/validations/NameValidation.scala [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+ * limitations under the License.
+ */
+
+package v1.controllers.requestParsers.validators.validations
+
+
+object NameValidation {
+
+  private val nameRegex =
+    "^[A-Za-z0-9 ]{1,}+$"
```
[the remark this answers is not in the document]
Comment (DreamcastKnight) [reply app/v1/controllers/requestParsers/validators/validations/NameValidation.scala:23]: I think this is another case where I'll need to rectify in spec, as the validation doesn't work without the space in the regex

File: app/v1/controllers/requestParsers/validators/AmendReliefInvestmentValidator.scala [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+    def generateFailedPaths[T <: Product](r: T): Seq[String] = {
+      val map: Map[String, Any] = toFieldNameMap(r)
+
+      map.collect {
+        case (objectName, list: List[T]) =>
+          list.zipWithIndex.flatMap {
+            case (innerObject, i) =>
+              toFieldNameMap(innerObject).collect {
+                case (fieldName, Some(value: BigDecimal)) => NumberValidation.validate(value, s"$objectName/[$i]/$fieldName")
```
Comment (james-work-account) [inline app/v1/controllers/requestParsers/validators/AmendReliefInvestmentValidator.scala:52]: remove square brackets (per refinement session) - update tests too

File: app/v1/controllers/requestParsers/validators/AmendReliefInvestmentValidator.scala [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+
+    def dateValidations[T <: Product](r: T): Seq[String] = {
+      val map: Map[String, Any] = toFieldNameMap(r)
+
+      map.collect {
+        case (objectName, list: List[T]) =>
+          list.zipWithIndex.flatMap {
+            case (innerObject, i) =>
+              toFieldNameMap(innerObject).collect {
+                case (fieldName: String, Some(value: String)) if fieldName.take(4) == "date" => DateValidation.validate(value, s"$objectName/[$i]/$fieldName")
```
Comment (james-work-account) [inline app/v1/controllers/requestParsers/validators/AmendReliefInvestmentValidator.scala:75]: remove square brackets (per refinement session) - update tests too

File: app/v1/controllers/requestParsers/validators/AmendReliefInvestmentValidator.scala [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+    def nameValidations[T <: Product](r: T): Seq[String] = {
+      val map: Map[String, Any] = toFieldNameMap(r)
+
+      map.collect {
+        case (objectName, list: List[T]) =>
+          list.zipWithIndex.flatMap {
+            case (innerObject, i) =>
+              toFieldNameMap(innerObject).collect {
+                case (fieldName: String, Some(value: String)) if fieldName.take(4) == "name" || fieldName.takeRight(4) == "Name" => NameValidation.validate(value, s"$objectName/[$i]/$fieldName")
```
Comment (james-work-account) [inline app/v1/controllers/requestParsers/validators/AmendReliefInvestmentValidator.scala:99]: remove square brackets (per refinement session) - update tests too

File: app/v1/controllers/requestParsers/validators/AmendReliefInvestmentValidator.scala [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+
+    def investmentRefValidations[T <: Product](r: T): Seq[String] = {
+      val map: Map[String, Any] = toFieldNameMap(r)
+
+      map.collect {
+        case (objectName, list: List[T]) =>
+          list.zipWithIndex.flatMap {
+            case (innerObject, i) =>
+              toFieldNameMap(innerObject).collect {
+                case (fieldName: String, Some(value: String)) if fieldName.takeRight(3) == "Ref" => InvestmentRefValidation.validate(value, s"$objectName/[$i]/$fieldName")
```
Comment (james-work-account) [inline app/v1/controllers/requestParsers/validators/AmendReliefInvestmentValidator.scala:123]: remove square brackets (per refinement session) - update tests too

**Pull request merged**